shell/browser/file_select_helper.cc

```diff
@@ -0,0 +1,244 @@
+#include <utility>
+#include <vector>
+
+#include "shell/browser/file_select_helper.h"
+
+#include "base/bind.h"
+#include "base/files/file_enumerator.h"
+#include "base/files/file_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "chrome/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "shell/browser/electron_browser_context.h"
+#include "shell/browser/javascript_environment.h"
+#include "shell/browser/ui/file_dialog.h"
+#include "shell/common/gin_converters/callback_converter.h"
+#include "shell/common/gin_converters/file_path_converter.h"
+
+using blink::mojom::FileChooserFileInfo;
+using blink::mojom::FileChooserFileInfoPtr;
+using blink::mojom::FileChooserParams;
+using blink::mojom::NativeFileInfo;
+
+namespace {
+void DeleteFiles(std::vector<base::FilePath> paths) {
+  for (auto& file_path : paths)
+    base::DeleteFile(file_path);
+}
+}  // namespace
+
+FileSelectHelper::FileSelectHelper(
+    content::RenderFrameHost* render_frame_host,
+    scoped_refptr<content::FileSelectListener> listener,
+    FileChooserParams::Mode mode)
+    : render_frame_host_(render_frame_host),
+      listener_(std::move(listener)),
+      mode_(mode) {
+  DCHECK(render_frame_host_);
+  DCHECK(listener_);
+
+  web_contents_ = content::WebContents::FromRenderFrameHost(render_frame_host);
+  DCHECK(web_contents_);
+  content::WebContentsObserver::Observe(web_contents_);
+}
+
+FileSelectHelper::~FileSelectHelper() = default;
+
+void FileSelectHelper::ShowOpenDialog(
+    const file_dialog::DialogSettings& settings) {
+  v8::Isolate* isolate = electron::JavascriptEnvironment::GetIsolate();
+  v8::HandleScope scope(isolate);
+  gin_helper::Promise<gin_helper::Dictionary> promise(isolate);
+
+  auto callback = base::BindOnce(&FileSelectHelper::OnOpenDialogDone, this);
+  ignore_result(promise.Then(std::move(callback)));
+
+  file_dialog::ShowOpenDialog(settings, std::move(promise));
+}
+
+void FileSelectHelper::ShowSaveDialog(
+    const file_dialog::DialogSettings& settings) {
+  v8::Isolate* isolate = electron::JavascriptEnvironment::GetIsolate();
+  v8::HandleScope scope(isolate);
+  gin_helper::Promise<gin_helper::Dictionary> promise(isolate);
+
+  auto callback = base::BindOnce(&FileSelectHelper::OnSaveDialogDone, this);
+  ignore_result(promise.Then(std::move(callback)));
+
+  file_dialog::ShowSaveDialog(settings, std::move(promise));
+}
+
+// net::DirectoryLister::DirectoryListerDelegate
+void FileSelectHelper::OnListFile(
+    const net::DirectoryLister::DirectoryListerData& data) {
+  // We don't want to return directory paths, only file paths
+  if (data.info.IsDirectory())
+    return;
+
+  lister_paths_.push_back(data.path);
+}
+
+void FileSelectHelper::RunSelectionEnd() {
+  // If there are temporary files, then this instance needs to stick around
+  // until web_contents_ is destroyed, so that this instance can delete the
+  // temporary files.
+  if (!temporary_files_.empty())
+    return;
```

Comment:
why does the instance have to wait until web_contents_ is destroyed, can't we do the following ?

```
if (!temporary_files_.empty()) {
  DeleteTemporaryFiles();
  Release();
}
```

Reply:
@deepak1556 the `temporary` is a bit of a misnomer - these files are the zipped files stored in the `zip_cache`. They should remain alive as long as the webcontents is alive, since that's likely how long the user would in some fashion need to interact with them